docs/model-dev-guide/apis-howto/deepspeed/autotuning.rst

```diff
@@ -0,0 +1,302 @@
+
+   hparams = self.context.get_hparams()
+   parser = HfArgumentParser(TrainingArguments)
+   args = sys.argv[1:]
+   args = dsat.get_hf_args_with_overwrites(args, hparams)
+   training_args = parser.parse_args_into_dataclasses(args, look_for_args_file=False)
+
+   det_callback = DetCallback(core_context, ...)
+   trainer = Trainer(args=training_args, ...)
+   with dsat.dsat_reporting_context(core_context, op=det_callback.current_op):
+       train_result = trainer.train(resume_from_checkpoint=checkpoint)
+
+.. important::
+
+   -  The ``dsat_reporting_context`` context manager shares the same initial
+      :class:`~determined.core.SearcherOperation` as the ``DetCallback`` instance through its
+      ``op=det_callback.current_op`` argument.
+
+   -  The entire ``train`` method of the HuggingFace trainer is now wrapped in the
+      ``dsat_reporting_context`` context manager.
+
+Examples which use DeepSpeed Autotune with HuggingFace Trainer can be found in the `Determined
+GitHub Repo <https://github.com/determined-ai/determined/tree/master/examples/hf_trainer_api>`__
+under ``examples/hf_trainer_api``.
+
+******************
+ Advanced Options
+******************
+
+The command-line entrypoint to ``dsat`` has various available options, some of them
+search-algorithm-specific. All available options for any given search method can be found as in
+
+.. code::
+
+   python3 -m determined.pytorch.dsat asha --help
+
+Below, we highlight particularly important flags and describe the search algorithms in some more
+detail.
+
+General Options
+===============
+
+The following options are available for every search method.
+
+-  ``--max-trials``: The maximum number of trials to run. Default: ``50``.
+
+-  ``--max-concurrent-trials``: The maximum number of trials that can run concurrently. Default:
+   ``16``.
+
+-  ``--max-slots``: The maximum number of slots that can be used concurrently. Defaults to ``None``,
+   i.e., there is no limit by default.
+
+-  ``--metric``: The metric to be optimized. Defaults to ``FLOPS-per-gpu``. Other available options
+   are ``throughput``, ``forward``, ``backward``, and ``latency``.
+
+-  ``--run-full-experiment``: If specified, after the ``dsat`` experiment has completed, a
+   ``single`` experiment will be launched using the specifications in the ``deepspeed.yaml``
+   overwritten with the best-found DS configuration parameters.
+
+-  ``--zero-stages``: This flag allows the user to limit the search to a subset of the stages by
+   providing a space-separated list, as in ``--zero-stages 2 3``. Default: ``1 2 3``.
+
+.. _asha-options:
+
+``asha`` Options
+================
+
+The ``asha`` search algorithm randomly generates various DeepSpeed configurations and attempts to
+tune the batch size for each such configuration through a binary search. ``asha`` adaptively
+allocates resources to explore each configuration (providing more resources to promising lineages)
```

Comment:
The ``asha`` search algorithm randomly generates various DeepSpeed configurations and attempts to
tune the batch size for each configuration through a binary search.